File: lib/core/permissions/permission_service.dart
/// App-specific permission types — decoupled from permission_handler package.
enum AppPermission {
  camera,
  photos,
  location,
  notifications,
  microphone,
  storage,
}

/// App-specific permission status — mirrors permission_handler statuses.
enum AppPermissionStatus {
  granted,
  denied,
  permanentlyDenied,
  restricted,
  limited,
}

/// Abstract contract for checking and requesting platform permissions.
abstract class PermissionService {
  /// Checks current status of [permission] without prompting the user.
  Future<AppPermissionStatus> check(AppPermission permission);

  /// Requests [permission] from the user; returns updated status.
  Future<AppPermissionStatus> request(AppPermission permission);

  /// Opens the app settings page for the user to change permissions manually.
  Future<bool> openSettings();
}

File: shared/src/commonMain/kotlin/dev/tsnanh/creativenote/core/permissions/PermissionService.kt
package dev.tsnanh.creativenote.core.permissions

enum class AppPermission { Camera, Photos, Location, Notifications }
enum class PermissionStatus { Granted, Denied, Limited, NotDetermined }

interface PermissionService {
    suspend fun status(permission: AppPermission): PermissionStatus
    suspend fun request(permission: AppPermission): PermissionStatus
}

class DefaultPermissionService : PermissionService {
    override suspend fun status(permission: AppPermission) = PermissionStatus.NotDetermined
    override suspend fun request(permission: AppPermission) = PermissionStatus.Denied
}
